pkg/virt-launcher/virtwrap/manager_cloudhv.go

```diff
@@ -0,0 +1,538 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Intel Corporation.
+ *
+ */
+
+package virtwrap
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"kubevirt.io/client-go/log"
+
+	cloudinit "kubevirt.io/kubevirt/pkg/cloud-init"
+	"kubevirt.io/kubevirt/pkg/config"
+	containerdisk "kubevirt.io/kubevirt/pkg/container-disk"
+	"kubevirt.io/kubevirt/pkg/emptydisk"
+	"kubevirt.io/kubevirt/pkg/hooks"
+	"kubevirt.io/kubevirt/pkg/network/cache"
+	netsetup "kubevirt.io/kubevirt/pkg/network/setup"
+	cmdclient "kubevirt.io/kubevirt/pkg/virt-handler/cmd-client"
+
+	v1 "kubevirt.io/api/core/v1"
+
+	ephemeraldisk "kubevirt.io/kubevirt/pkg/ephemeral-disk"
+	cmdv1 "kubevirt.io/kubevirt/pkg/handler-launcher-com/cmd/v1"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/converter"
+	openapiClient "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/openapi/cloud-hypervisor/client"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/stats"
+)
+
+const (
+	clientTimeout = 300
+)
+
+type CloudHvDomainManager struct {
+	client               *openapiClient.DefaultApiService
+	vmConfig             openapiClient.VmConfig
+	ephemeralDiskCreator ephemeraldisk.EphemeralDiskCreatorInterface
+	started              bool
+	domain               *api.Domain
+}
+
+func NewCloudHvDomainManager(apiSocketPath, ephemeralDiskDir, efiDir string, ephemeralDiskCreator ephemeraldisk.EphemeralDiskCreatorInterface) (*CloudHvDomainManager, error) {
+	return newCloudHvDomainManager(apiSocketPath, ephemeralDiskDir, efiDir, ephemeralDiskCreator)
+}
+
+func newCloudHvDomainManager(apiSocketPath, ephemeralDiskDir, efiDir string, ephemeralDiskCreator ephemeraldisk.EphemeralDiskCreatorInterface) (*CloudHvDomainManager, error) {
+	apiSocketAddr, err := net.ResolveUnixAddr("unix", apiSocketPath)
+	if err != nil {
+		return nil, err
+	}
+
+	clientConfig := openapiClient.NewConfiguration()
+	clientConfig.HTTPClient = &http.Client{
+		Transport: &http.Transport{
+			DialContext: func(ctx context.Context, network, path string) (net.Conn, error) {
+				return net.DialUnix("unix", nil, apiSocketAddr)
+			},
+		},
+	}
+
+	return &CloudHvDomainManager{
+		client: openapiClient.NewAPIClient(clientConfig).DefaultApi,
+		vmConfig: openapiClient.VmConfig{
+			Kernel: openapiClient.KernelConfig{
+				Path: filepath.Join(efiDir, "CLOUDHV.fd"),
+			},
+		},
+		ephemeralDiskCreator: ephemeralDiskCreator,
+		started:              false,
+		domain:               &api.Domain{},
```

Comment:
I see that you got this running with only sending back an extremely minimal `api.Domain` via the notify channel. I think this is one of the places where we want to look closer and identify which pieces we really need and come up with a more focused representation which get's sent back.

On the other hand, the API may be fine as is and focusing more on being able to report on what a hypervisor supports may be more interesting. The nice part about this is, that independent on how cloud-hypervisor would be used, we need to make kubevirt adapt to a different set of supported features (api, controller and handler). Does not have to be over-complex. It is probably enough to just let every component know the hypervisor type. Then all components are shipped anyway as one endity like for libvirt/qemu.

Reply:
`api.Domain` is quite complex but it's meant to support the most complex use case which is libvirt/qemu. I don't mind having Cloud Hypervisor using this structure as the API to carry information upward, but one thing that might be nice for possible future integrations would be to document the minimal amount of things that are expected to be filled through this API.
For instance, I spent a bit of time debugging the `Pause` use case because I didn't understand why the VMI wasn't considered `paused`. I found out the code had some specific expectation about the status `Paused` and the associated `Reason` (I wasn't  providing a `Reason`, which led to the unexpected behavior).